filebeat/harvester/reader/line.go

```diff
@@ -24,55 +24,34 @@ type Line struct {
 	decoder    transform.Transformer
 }
 
-func NewLine(
-	input io.Reader,
-	codec encoding.Encoding,
-	bufferSize int,
-) (*Line, error) {
-	l := &Line{}
-
-	if err := l.init(input, codec, bufferSize); err != nil {
-		return nil, err
-	}
+// NewLine creates a new Line reader object
+func NewLine(input io.Reader, codec encoding.Encoding, bufferSize int) (*Line, error) {
 
-	return l, nil
-}
+	encoder := codec.NewEncoder()
 
-func (l *Line) init(
-	reader io.Reader,
-	codec encoding.Encoding,
-	bufferSize int,
-) error {
-	l.reader = reader
-	l.codec = codec
-	l.bufferSize = bufferSize
-
-	l.codec.NewEncoder()
-	nl, _, err := transform.Bytes(l.codec.NewEncoder(), []byte{'\n'})
+	// Create newline char based on encoding
+	nl, _, err := transform.Bytes(encoder, []byte{'\n'})
 	if err != nil {
-		return err
+		return nil, err
 	}
 
-	l.nl = nl
-	l.decoder = l.codec.NewDecoder()
-	l.inBuffer = streambuf.New(nil)
-	l.outBuffer = streambuf.New(nil)
-	return nil
+	return &Line{
+		reader:     input,
+		codec:      codec,
+		bufferSize: bufferSize,
+		nl:         nl,
+		decoder:    codec.NewDecoder(),
+		inBuffer:   streambuf.New(nil),
+		outBuffer:  streambuf.New(nil),
+	}, nil
 }
 
 func (l *Line) Next() ([]byte, int, error) {
-	for {
-		// read next 'potential' line from input buffer/reader
-		err := l.advance()
-		if err != nil {
-			return nil, 0, err
-		}
 
-		// check last decoded byte really being '\n'
-		buf := l.outBuffer.Bytes()
-		if buf[len(buf)-1] == '\n' {
```

Comment:
tl;dr It can be perfectly ok advance thinking it found a newline, but in utf8 actually not being a newline due to relying on byte comparisons only.

the reason it has been a loop was: what if encoded newline was found due to being a prefix-/suffix-pattern used in encoding another symbol. The check for newline in both places is based on byte comparisons not taking the actual encoding into account. Checking last symbol in array is actually a newline character would require you to iterate array rune by rune (not byte by byte), taking the current encoding into account.

This check mostly acts as second hopefully-we-catch-this-somehow-heuristics in case of false-positive in `advance`.